torch/csrc/jit/script/sugared_value.cpp

```diff
@@ -130,6 +130,80 @@ std::vector<std::shared_ptr<SugaredValue>> SimpleValue::asTuple(
                          << " cannot be used as a tuple";
 }
 
+void UserTypeValue::assign(
+    const SourceRange& loc,
+    Method& m,
+    const std::string& field,
+    Value* newValue) {
+  auto expectedType = type_->getAttribute(field);
+  if (!expectedType) {
+    // We don't have an attribute with this name, either add it to the type
+    // definition or throw an error
+    if (m.name() == "__init__") {
+      type_->addAttribute(field, newValue->type());
+      expectedType = newValue->type();
+    } else {
+      throw ErrorReport(loc)
+          << "Tried to assign to nonexistent attribute " << field
+          << ". Did you forget to initialize it in __init__()?";
+    }
+  }
+
+  // Check type correctness
+  const auto newType = newValue->type();
+  if (!newType->isSubtypeOf(expectedType)) {
+    throw ErrorReport(loc) << "Wrong type for attribute assignment. Expected "
+                           << expectedType->str() << " but got "
+                           << newType->str();
+  }
+
+  auto& g = *m.graph();
+  g.insertNode(g.createSetAttr(value_, field, newValue));
+}
+
+std::shared_ptr<SugaredValue> UserTypeValue::attr(
+    const SourceRange& loc,
+    Method& m,
+    const std::string& field) {
+  AT_ASSERT(value_);
+  if (auto method = type_->module()->find_method(field)) {
+    return std::make_shared<MethodValue>(shared_from_this(), *method);
+  }
+
+  if (!type_->hasAttribute(field)) {
+    throw ErrorReport(loc)
+        << "Tried to access to nonexistent attribute " << field
+        << ". Did you forget to initialize it in __init__()?";
+  }
+  auto& g = *m.graph();
+  auto n = g.insertNode(g.createGetAttr(value_, field));
+  return std::make_shared<SimpleValue>(n->output());
+}
+
+std::shared_ptr<SugaredValue> UserTypeValue::call(
```

Comment:
This doesn't seem right, this would do:

```
a = <some user type>
b = a() # constructs a new instance of the type of a?
```

Instead we probably want to make a different SugaredValue that represents the class itself which has this logic in it.